### PR TITLE
Fix Interrupt message specs

### PR DESCRIPTION
**Pull request by bjfish**

### CODE REVIEW

File: src/main/ruby/core/exception.rb [hunk 1]
```diff
@@ -390,37 +401,56 @@ class KeyError < IndexError
...
-  def initialize(signo = nil, signm = nil)
-    # MRI overrides this behavior just for SignalException itself
-    # but not for anything that inherits from it, therefore we
-    # need this ugly check to make sure it works as intended.
-    return super(signo) unless self.class == SignalException
-    if signo.is_a? Integer
-      unless @signm = Signal::Numbers[signo]
-        raise ArgumentError, "invalid signal number #{signo}"
+
+  def initialize(*args)
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: Can we use a simpler signature here like `sig, message=nil`?

File: src/main/ruby/core/exception.rb [hunk 2]
```diff
@@ -390,37 +401,56 @@ class KeyError < IndexError
...
-      @signo = signo
-      @signm = signm || "SIG#{@signm}"
-    elsif signo
-      if signm
-        raise ArgumentError, 'wrong number of arguments (2 for 1)'
+    end
+    Rubinius::Type.check_arity(args.size, 1, argnum)
+    if argnum == 2
+      signo = Rubinius::Type.rb_num2int(sig)
+      if signo < 0 || signo > (Signal::Names.values.max + 1)
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: The max Signal value should be computed only once.

Comment (eregon) [reply src/main/ruby/core/exception.rb]: Isn't it enough to just check if `signo` is in `Signal::Numbers`?

File: src/main/ruby/core/exception.rb (hunk 1 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: I picked this signature for compatibility.

Comment (eregon) [reply src/main/ruby/core/exception.rb]: What compatibility?
The value of `arity`? Is that spec'd? I'd consider it an unimportant implementation detail of MRI in general.

Comment (eregon) [reply src/main/ruby/core/exception.rb]: (MRI cannot handle optional arguments without changing the arity to -1 in methods defined in C iirc)

Comment (bjfish) [reply src/main/ruby/core/exception.rb]: @eregon Updated, I used `initialize(sig, message = undefined)` because the behavior appears to vary by the number of arguments.
```
irb(main):003:0> SignalException.new(2).signm
=> "SIGINT"
irb(main):004:0> SignalException.new(2, nil).signm
```

File: src/main/ruby/core/exception.rb (hunk 2 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: Fixed, to use `Signal::Numbers`

File: src/main/ruby/core/exception.rb [hunk 3]
```diff
@@ -236,6 +236,17 @@ class NotImplementedError < ScriptError
...
 class Interrupt < SignalException
+  def initialize(*args)
+    case args.size
+    when 0
+      message = nil
+    when 1
+      message = args.first
+    else
+      raise ArgumentError, "wrong number of arguments (#{args.size} for 0..1)"
+    end
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: Can you also use required/optional arguments here instead of manually checking args.size?

File: src/main/ruby/core/exception.rb [hunk 4]
```diff
@@ -390,37 +401,44 @@ class KeyError < IndexError
...
+        sig = sig.to_s
+      else
+        sig = StringValue(sig)
       end
-      @signo = signo
-      @signm = signm || "SIG#{@signm}"
-    elsif signo
-      if signm
-        raise ArgumentError, 'wrong number of arguments (2 for 1)'
+      signm = sig
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: Could you give a meaningful name to this variable, like `signame` or `name` or `signal_name`?

File: src/main/ruby/core/exception.rb [hunk 5]
```diff
@@ -390,37 +401,44 @@ class KeyError < IndexError
...
+      signo = Signal::Names[signm]
+      if signo.nil?
+        raise ArgumentError, "invalid signal name SIG#{sig}"
       end
-      signm = signm[3..-1] if signm.start_with? 'SIG'
-      unless @signo = Signal::Names[signm]
-        raise ArgumentError, "invalid signal name #{signm}"
+      sig = 'SIG%s' % signm
+    else
+      if signo < 0 || !Signal::Numbers.has_key?(signo)
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: `has_key?` is deprecated iirc, use `key?` instead.

Comment (eregon) [reply src/main/ruby/core/exception.rb]: Since Signal::Numbers doesn't have negative keys, the second part of the check is enough.

File: src/main/ruby/core/exception.rb [hunk 6]
```diff
@@ -390,37 +401,44 @@ class KeyError < IndexError
...
-      else
-        signm = StringValue(signm)
+      signo = Signal::Names[signm]
+      if signo.nil?
+        raise ArgumentError, "invalid signal name SIG#{sig}"
       end
-      signm = signm[3..-1] if signm.start_with? 'SIG'
-      unless @signo = Signal::Names[signm]
-        raise ArgumentError, "invalid signal name #{signm}"
+      sig = 'SIG%s' % signm
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: Better variable name: `name_with_prefix`

File: src/main/ruby/core/exception.rb [hunk 7]
```diff
@@ -390,37 +401,44 @@ class KeyError < IndexError
...
+    else
+      if signo < 0 || !Signal::Numbers.has_key?(signo)
+        raise ArgumentError, "invalid signal number (#{signo})"
       end
-      @signm = "SIG#{signm}"
+      sig = if undefined.equal?(message)
+              s = Signal::Numbers[signo]
+              if s.nil?
+                'SIG%d' % signo
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: This case is no longer possible, is it?
Since we check that signo is in Signal::Numbers above.

File: src/main/ruby/core/exception.rb (hunk 6 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: Fixed.

File: src/main/ruby/core/exception.rb (hunk 3 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: Yes, fixed.

File: src/main/ruby/core/exception.rb (hunk 4 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: Yes, updated to `signal_name`

File: src/main/ruby/core/exception.rb (hunk 7 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: I found how this is possible, for example on macOS:
```
irb(main):002:0> SignalException.new(32)
=> #<SignalException: SIG32>
```
I think `SignalException.new(0)` on Linux might do this too.

File: src/main/ruby/core/exception.rb (hunk 5 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: I've updated this to check `signo > Signal::NSIG` for the case mentioned here:
https://github.com/graalvm/truffleruby/pull/819#discussion_r155702914

File: src/main/ruby/core/signal.rb [hunk 8]
```diff
@@ -120,6 +120,8 @@ def self.trap(sig, prc=nil, &block)
     old ? old : nil
   end
 
+  NSIG = Numbers.keys.max + 1
```
Comment (eregon) [inline src/main/ruby/core/signal.rb]: Can you define this higher next to other constants (~line 50)?
Also, it could either compute the max as part of the existing loop, or use `NSIG` from the Rubinius config (would need to be added, I can help with that).

File: src/main/ruby/core/exception.rb [hunk 9]
```diff
@@ -390,37 +393,44 @@ class KeyError < IndexError
...
-      unless @signo = Signal::Names[signm]
-        raise ArgumentError, "invalid signal name #{signm}"
+      name_with_prefix = 'SIG%s' % signal_name
+    else
+      if signo < 0 || signo > Signal::NSIG
+        raise ArgumentError, "invalid signal number (#{signo})"
       end
-      @signm = "SIG#{signm}"
+      name_with_prefix = if undefined.equal?(message)
+                           s = Signal::Numbers[signo]
```
Comment (eregon) [inline src/main/ruby/core/exception.rb]: Could be `name` instead of `s` :)

File: src/main/ruby/core/exception.rb (hunk 7 above)
Comment (eregon) [reply src/main/ruby/core/exception.rb]: I see, 32 also shows the same on Linux.
This sounds very much of a corner case, but given it seems easy to implement, it sounds good.
It would be good to have a spec or test for it though, unless there is already a spec/test covering this.

File: src/main/ruby/core/exception.rb (hunk 9 above)
Comment (bjfish) [reply src/main/ruby/core/exception.rb]: Updated to `name`.

File: src/main/ruby/core/signal.rb (hunk 8 above)
Comment (bjfish) [reply src/main/ruby/core/signal.rb]: Moved this ~line 50.

**Comment by bjfish** [comment]
@eregon Thanks, I rebased to master to de-conflict .
